### PR TITLE
Fix Memory Leak in GLUE Trainer

### PR DESCRIPTION
**Pull request by moinnadeem**
Cool, I'm a bit late on this, sorry for the delay. I thought I had fixed the memory leak, but turns out I accidentally introduced a deadlock.

I think the core of the memory leak might actually reside with Composer or how we are creating the dataloaders, and could relate to the zombie processes that everyone else is seeing. However, our GLUE Trainer also wasn't killing each Trainer instance when it was finished training, but instead was re-using the process to launch further trainers. Therefore, the core of this solution relies on fully killing the process before spawning a new one (setting `maxtasksperchild = 1`). 

Finally, we also need to turn off multiple workers for the dataloader due to multiprocessing, and convert the Torch tensors to primitives in order to be sent over multiprocessing.

### CODE REVIEW

**Review by alextrott16 — COMMENTED** [review]
Brilliant! Thanks, Moin!

I'm not familiar with the inner workings of the `pool` object, and why you want to use `apply_async` instead of the `submit` method, so I can't comment on that choice.

My only question is how deeply has this been tested? If you can confirm that it runs without issue, then I'm happy to approve. Otherwise, let me know if you need my help in performing more tests.

**Comment by moinnadeem** [comment]
@alextrott16 I have tested it with setting `task_hparams.train_subset_num_batches = 100` and seeing things train to completion without any memory leaks. I am running a test now on full fine-tuning: one hour in and memory is constant, compared to linear growth without the fix.

However, if you want to help test, I surely wouldn't complain. I am reasonably confident that this isn't memory leaking, but I have some slight worries that something might go wrong somewhere else (just out of paranoia more than anything else)

File: examples/glue/run_glue_trainer.py [hunk 1]
```diff
@@ -330,8 +368,30 @@ def train_finetune(
 
     trainer.fit()
     print(f'\nFINISHED TRAINING TASK {task.upper()}\n')
+
+    cpu_metrics = DeviceCPU().batch_to_device(trainer.state.current_metrics)
+    # convert the Torch tensors to primitives so they can be sent over multiprocessing
+    cpu_metrics = _convert_torch_tensor_to_primitives(cpu_metrics)
```
Comment (alextrott16) [inline examples/glue/run_glue_trainer.py:375]: Please add a `trainer.close()` call before this function exits

**Review by alextrott16 — CHANGES_REQUESTED** [review]

File: examples/glue/run_glue_trainer.py (hunk 1 above)
Comment (moinnadeem) [reply examples/glue/run_glue_trainer.py:375]: Done!

**Review by alextrott16 — APPROVED** [review]
LGTM, but please hold off on merging it until the testing is complete. Will follow up tomorrow!

**Comment by alextrott16** [comment]
The fine-tuning test passed with flying colors!

**Pull request merged**